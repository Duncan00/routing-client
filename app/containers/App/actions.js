import {
  LOAD_ROUTE,
  LOAD_ROUTE_SUCCESS,
  LOAD_ROUTE_ERROR,
  CREATE_ROUTE,
  CREATE_ROUTE_SUCCESS,
  CREATE_ROUTE_ERROR,
  RESET,
} from './constants';

export function loadRoute() {
  return {
    type: LOAD_ROUTE,
  };
}

export function routeLoaded(data) {
  return {
    type: LOAD_ROUTE_SUCCESS,
    data,
  };
}

export function routeLoadingError(error) {
  return {
    type: LOAD_ROUTE_ERROR,
    error,
  };
}

export function createRoute(startingLocation, dropOffPoint) {
  return {
    type: CREATE_ROUTE,
    startingLocation,
    dropOffPoint,
  };
}

export function routeCreated(data) {
  return {
    type: CREATE_ROUTE_SUCCESS,
    data,
  };
}

export function routeCreatingError(error) {
  return {
    type: CREATE_ROUTE_ERROR,
    error,
  };
}

export function reset() {
  return {
    type: RESET,
  };
}
