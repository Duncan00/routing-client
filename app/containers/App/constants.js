/*
 * AppConstants
 * Each action has a corresponding type, which the reducer knows and picks up on.
 * To avoid weird typos between the reducer and the actions, we save them as
 * constants here. We prefix them with 'yourproject/YourComponent' so we avoid
 * reducers accidentally picking up actions they shouldn't.
 *
 * Follow this format:
 * export const YOUR_ACTION_CONSTANT = 'yourproject/YourContainer/YOUR_ACTION_CONSTANT';
 */

export const LOAD_ROUTE = 'routing-client/App/LOAD_ROUTE';
export const LOAD_ROUTE_SUCCESS = 'routing-client/App/LOAD_ROUTE_SUCCESS';
export const LOAD_ROUTE_ERROR = 'routing-client/App/LOAD_ROUTE_ERROR';

export const CREATE_ROUTE = 'routing-client/App/CREATE_ROUTE';
export const CREATE_ROUTE_SUCCESS = 'routing-client/App/CREATE_ROUTE_SUCCESS';
export const CREATE_ROUTE_ERROR = 'routing-client/App/CREATE_ROUTE_ERROR';

export const RESET = 'routing-client/App/RESET';
