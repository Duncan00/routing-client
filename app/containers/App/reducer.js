import produce from 'immer';
import { pick } from 'lodash';
import {
  LOAD_ROUTE,
  LOAD_ROUTE_SUCCESS,
  LOAD_ROUTE_ERROR,
  CREATE_ROUTE,
  CREATE_ROUTE_SUCCESS,
  CREATE_ROUTE_ERROR,
  RESET,
} from './constants';

// The initial state of the App
export const initialState = {
  loading: false,
  error: false,
  api: process.env.API,
  user_data: {
    route_token: '',
    route: false,
  },
};

/* eslint-disable default-case, no-param-reassign */
const appReducer = (state = initialState, action) =>
  // eslint-disable-next-line consistent-return
  produce(state, draft => {
    switch (action.type) {
      case LOAD_ROUTE:
        draft.loading = true;
        draft.error = false;
        draft.user_data.route = false;
        break;

      case LOAD_ROUTE_SUCCESS:
        draft.user_data.route = pick(action.data, [
          'path',
          'total_distance',
          'total_time',
        ]);
        draft.loading = false;
        break;

      case LOAD_ROUTE_ERROR:
        draft.error = action.error.message;
        draft.loading = false;
        break;

      case CREATE_ROUTE:
        draft.loading = true;
        draft.error = false;
        draft.user_data.route_token = '';
        break;

      case CREATE_ROUTE_SUCCESS: {
        draft.loading = false;
        draft.user_data.route_token = action.data.token;
        break;
      }

      case CREATE_ROUTE_ERROR:
        draft.loading = false;
        draft.error = action.error.message;
        break;

      case RESET:
        return initialState;
    }
  });

export default appReducer;
