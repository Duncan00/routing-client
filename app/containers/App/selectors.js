import { createSelector } from 'reselect';
import { initialState } from './reducer';

const selectGlobal = state => state.global || initialState;

const makeSelectApi = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.api,
  );

const makeSelectRouteToken = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.user_data.route_token,
  );

const makeSelectRoute = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.user_data.route,
  );

const makeSelectError = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.error,
  );

const makeSelectLoading = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.loading,
  );

export {
  makeSelectApi,
  makeSelectRouteToken,
  makeSelectRoute,
  makeSelectError,
  makeSelectLoading,
};
