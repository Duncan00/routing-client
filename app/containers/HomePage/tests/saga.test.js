import { put, call, takeLatest } from 'redux-saga/effects';

import { CREATE_ROUTE, LOAD_ROUTE } from 'containers/App/constants';
import {
  routeLoaded,
  routeLoadingError,
  routeCreated,
  routeCreatingError,
} from 'containers/App/actions';

import routeData, { getRoute, createRoute, STATUS } from '../saga';

describe('getRoute Saga', () => {
  let getRouteGenerator;

  function getMockResponse(status) {
    return {
      status,
    };
  }

  beforeEach(() => {
    getRouteGenerator = getRoute();

    const selectApiDescriptor = getRouteGenerator.next().value;
    expect(selectApiDescriptor).toMatchSnapshot();

    const api = 'api.example.com';
    const selectTokenDescriptor = getRouteGenerator.next(api).value;
    expect(selectTokenDescriptor).toMatchSnapshot();

    const token = 'token';
    const callDescriptor = getRouteGenerator.next(token).value;
    expect(callDescriptor).toMatchSnapshot();
  });

  describe("When status === 'success' in response", () => {
    it('should put routeLoaded', () => {
      const response = getMockResponse(STATUS.SUCCESS);
      const putDescriptor = getRouteGenerator.next(response).value;
      expect(putDescriptor).toEqual(put(routeLoaded(response)));
    });
  });

  describe("When status === 'in progress' in response", () => {
    it("should retry call and put routeLoaded if status === 'success' in new response", () => {
      const inProgressResponse = getMockResponse(STATUS.IN_PROGRESS);
      const callDescriptor = getRouteGenerator.next(inProgressResponse).value;
      expect(callDescriptor).toMatchSnapshot();

      const successResponse = getMockResponse(STATUS.SUCCESS);
      const retryPutDescriptor = getRouteGenerator.next(successResponse).value;
      expect(retryPutDescriptor).toEqual(put(routeLoaded(successResponse)));
    });

    it("should throw error if status === 'in progress' in all 10 attempt responses", () => {
      const inProgressResponse = getMockResponse(STATUS.IN_PROGRESS);

      // Retry 9 times, total attempt 10 times
      for (let i = 0; i < 9; i += 1) {
        const callDescriptor = getRouteGenerator.next(inProgressResponse).value;
        expect(callDescriptor).toMatchSnapshot();
      }

      const putDescriptor = getRouteGenerator.next(inProgressResponse).value;
      expect(putDescriptor).toEqual(
        put(routeLoadingError(new Error('Server Busy. Please try again.'))),
      );
    });
  });

  describe("When status === 'failure' in response", () => {
    it('should put routeLoadingError', () => {
      const failureResponse = {
        ...getMockResponse(STATUS.FAILURE),
        error: 'Some errors',
      };
      const putDescriptor = getRouteGenerator.next(failureResponse).value;
      expect(putDescriptor).toEqual(
        put(routeLoadingError(new Error(failureResponse.error))),
      );
    });
  });
});

describe('createRoute Saga', () => {
  let createRouteGenerator;

  beforeEach(() => {
    createRouteGenerator = createRoute({
      startingLocation: 'starting_location',
      dropOffPoint: 'drop_off_point',
    });

    const selectApiDescriptor = createRouteGenerator.next().value;
    expect(selectApiDescriptor).toMatchSnapshot();

    const api = 'api.example.com';
    const callDescriptor = createRouteGenerator.next(api).value;
    expect(callDescriptor).toMatchSnapshot();
  });

  describe('When it requests the data successfully', () => {
    it('should put routeLoaded action and call getRoute', () => {
      const response = {};
      const putDescriptor = createRouteGenerator.next(response).value;
      expect(putDescriptor).toEqual(put(routeCreated(response)));

      const callDescriptor = createRouteGenerator.next().value;
      expect(callDescriptor).toEqual(call(getRoute));
    });
  });

  describe('When it responses errors', () => {
    it('should call routeLoadingError action', () => {
      const response = new Error('Some error');
      const putDescriptor = createRouteGenerator.throw(response).value;
      expect(putDescriptor).toEqual(put(routeCreatingError(response)));
    });
  });
});

describe('routeDataSaga', () => {
  const routeDataSaga = routeData();

  it('should start task to watch for CREATE_ROUTE actions', () => {
    const takeLatestLoadDescriptor = routeDataSaga.next().value;
    expect(takeLatestLoadDescriptor).toEqual(
      takeLatest(CREATE_ROUTE, createRoute),
    );
  });

  it('should start task to watch for LOAD_ROUTE actions', () => {
    const takeLatestVoteDescriptor = routeDataSaga.next().value;
    expect(takeLatestVoteDescriptor).toEqual(takeLatest(LOAD_ROUTE, getRoute));
  });
});
