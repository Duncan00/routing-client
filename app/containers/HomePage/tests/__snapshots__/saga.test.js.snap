// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createRoute Saga When it requests the data successfully should put routeLoaded action and call getRoute 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`createRoute Saga When it requests the data successfully should put routeLoaded action and call getRoute 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "api.example.com/mock/route/success",
      Object {
        "body": "{\\"origin\\":\\"starting_location\\",\\"destination\\":\\"drop_off_point\\"}",
        "headers": Object {
          "Content-Type": "application/json",
        },
        "method": "POST",
      },
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`createRoute Saga When it responses errors should call routeLoadingError action 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`createRoute Saga When it responses errors should call routeLoadingError action 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "api.example.com/mock/route/success",
      Object {
        "body": "{\\"origin\\":\\"starting_location\\",\\"destination\\":\\"drop_off_point\\"}",
        "headers": Object {
          "Content-Type": "application/json",
        },
        "method": "POST",
      },
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getRoute Saga When status === 'failure' in response should put routeLoadingError 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`getRoute Saga When status === 'failure' in response should put routeLoadingError 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`getRoute Saga When status === 'failure' in response should put routeLoadingError 3`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "api.example.com/route/token",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getRoute Saga When status === 'in progress' in response should retry call and put routeLoaded if status === 'success' in new response 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`getRoute Saga When status === 'in progress' in response should retry call and put routeLoaded if status === 'success' in new response 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`getRoute Saga When status === 'in progress' in response should retry call and put routeLoaded if status === 'success' in new response 3`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "api.example.com/route/token",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getRoute Saga When status === 'in progress' in response should retry call and put routeLoaded if status === 'success' in new response 4`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "api.example.com/route/token",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getRoute Saga When status === 'in progress' in response should throw error if status === 'in progress' in all 10 attempt responses 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`getRoute Saga When status === 'in progress' in response should throw error if status === 'in progress' in all 10 attempt responses 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`getRoute Saga When status === 'in progress' in response should throw error if status === 'in progress' in all 10 attempt responses 3`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "api.example.com/route/token",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getRoute Saga When status === 'in progress' in response should throw error if status === 'in progress' in all 10 attempt responses 4`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "api.example.com/route/token",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getRoute Saga When status === 'in progress' in response should throw error if status === 'in progress' in all 10 attempt responses 5`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "api.example.com/route/token",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getRoute Saga When status === 'in progress' in response should throw error if status === 'in progress' in all 10 attempt responses 6`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "api.example.com/route/token",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getRoute Saga When status === 'in progress' in response should throw error if status === 'in progress' in all 10 attempt responses 7`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "api.example.com/route/token",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getRoute Saga When status === 'in progress' in response should throw error if status === 'in progress' in all 10 attempt responses 8`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "api.example.com/route/token",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getRoute Saga When status === 'in progress' in response should throw error if status === 'in progress' in all 10 attempt responses 9`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "api.example.com/route/token",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getRoute Saga When status === 'in progress' in response should throw error if status === 'in progress' in all 10 attempt responses 10`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "api.example.com/route/token",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getRoute Saga When status === 'in progress' in response should throw error if status === 'in progress' in all 10 attempt responses 11`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "api.example.com/route/token",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getRoute Saga When status === 'in progress' in response should throw error if status === 'in progress' in all 10 attempt responses 12`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "api.example.com/route/token",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getRoute Saga When status === 'success' in response should put routeLoaded 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`getRoute Saga When status === 'success' in response should put routeLoaded 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`getRoute Saga When status === 'success' in response should put routeLoaded 3`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "api.example.com/route/token",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;
